File: stripe/src/main/java/com/stripe/android/Stripe3DS2AuthParams.java
package com.stripe.android;

import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Stripe3DS2AuthParams {

    static final String FIELD_APP = "app";
    static final String FIELD_SOURCE = "source";

    private static final String FIELD_SDK_APP_ID = "sdkAppID";
    private static final String FIELD_SDK_TRANS_ID = "sdkTransID";
    private static final String FIELD_SDK_ENC_DATA = "sdkEncData";
    private static final String FIELD_SDK_EPHEM_PUB_KEY = "sdkEphemPubKey";
    private static final String FIELD_SDK_MAX_TIMEOUT = "sdkMaxTimeout";
    private static final String FIELD_MESSAGE_VERSION = "messageVersion";
    private static final String FIELD_DEVICE_RENDER_OPTIONS = "deviceRenderOptions";

    private static final String FIELD_SDK_INTERFACE = "sdkInterface";
    private static final String FIELD_SDK_UI_TYPE = "sdkUiType";

    @NonNull private final String mSourceId;
    @NonNull private final String mDeviceData;
    @NonNull private final String mSdkAppId;
    @NonNull private final String mSdkTransactionId;
    @NonNull private final String mSdkEphemeralPublicKey;
    @NonNull private final String mMessageVersion;
    private final int mMaxTimeout;

    @VisibleForTesting
    Stripe3DS2AuthParams(@NonNull String sourceId,
                         @NonNull String sdkAppId,
                         @NonNull String sdkTransactionId,
                         @NonNull String deviceData,
                         @NonNull String sdkEphemeralPublicKey,
                         @NonNull String messageVersion,
                         int maxTimeout) {
        mSourceId = sourceId;
        mSdkAppId = sdkAppId;
        mDeviceData = deviceData;
        mSdkTransactionId = sdkTransactionId;
        mSdkEphemeralPublicKey = sdkEphemeralPublicKey;
        mMessageVersion = messageVersion;
        mMaxTimeout = maxTimeout;
    }

    @NonNull
    Map<String, Object> toParamMap() {
        final Map<String, Object> params = new HashMap<>();
        params.put(FIELD_SOURCE, mSourceId);
        params.put(FIELD_APP, createAppParams().toString());
        return params;
    }

    @NonNull
    private JSONObject createAppParams() {
        final JSONObject appParams = new JSONObject();
        try {
            appParams.put(FIELD_SDK_APP_ID, mSdkAppId);
            appParams.put(FIELD_SDK_TRANS_ID, mSdkTransactionId);
            appParams.put(FIELD_SDK_ENC_DATA, mDeviceData);
            appParams.put(FIELD_SDK_EPHEM_PUB_KEY, new JSONObject(mSdkEphemeralPublicKey));
            appParams.put(FIELD_SDK_MAX_TIMEOUT, mMaxTimeout);
            appParams.put(FIELD_MESSAGE_VERSION, mMessageVersion);
            appParams.put(FIELD_DEVICE_RENDER_OPTIONS, createDeviceRenderOptions());
        } catch (JSONException ignore) { }

        return appParams;
    }

    @NonNull
    private JSONObject createDeviceRenderOptions() {
        final JSONObject deviceRenderOptions = new JSONObject();
        try {
            deviceRenderOptions.put(FIELD_SDK_INTERFACE, "03");
            deviceRenderOptions.put(FIELD_SDK_UI_TYPE,
                    new JSONArray(Arrays.asList("01", "02", "03", "04", "05")));
        } catch (JSONException ignore) { }
        return deviceRenderOptions;
    }
}

File: stripe/src/main/java/com/stripe/android/Stripe3ds2AuthParams.kt
package com.stripe.android

import android.os.Parcelable
import com.stripe.android.model.StripeParamsModel
import java.text.DecimalFormat
import kotlinx.android.parcel.Parcelize
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

@Parcelize
internal data class Stripe3ds2AuthParams(
    private val sourceId: String,
    private val sdkAppId: String,
    private val sdkReferenceNumber: String,
    private val sdkTransactionId: String,
    private val deviceData: String,
    private val sdkEphemeralPublicKey: String,
    private val messageVersion: String,
    private val maxTimeout: Int,
    private val returnUrl: String?
) : StripeParamsModel, Parcelable {

    override fun toParamMap(): Map<String, Any> {
        val params = mapOf(
            FIELD_SOURCE to sourceId,
            FIELD_APP to createAppParams().toString()
        )

        return returnUrl?.let {
            params.plus(FIELD_FALLBACK_RETURN_URL to it)
        } ?: params
    }

    private fun createAppParams(): JSONObject {
        return try {
            JSONObject()
                .put(FIELD_SDK_APP_ID, sdkAppId)
                .put(FIELD_SDK_TRANS_ID, sdkTransactionId)
                .put(FIELD_SDK_ENC_DATA, deviceData)
                .put(FIELD_SDK_EPHEM_PUB_KEY, JSONObject(sdkEphemeralPublicKey))
                .put(FIELD_SDK_MAX_TIMEOUT, MAX_TIMEOUT_FORMATTER.format(maxTimeout.toLong()))
                .put(FIELD_SDK_REFERENCE_NUMBER, sdkReferenceNumber)
                .put(FIELD_MESSAGE_VERSION, messageVersion)
                .put(FIELD_DEVICE_RENDER_OPTIONS, createDeviceRenderOptions())
        } catch (ignore: JSONException) {
            JSONObject()
        }
    }

    private fun createDeviceRenderOptions(): JSONObject {
        return try {
            JSONObject()
                .put(FIELD_SDK_INTERFACE, "03")
                .put(FIELD_SDK_UI_TYPE, JSONArray(listOf("01", "02", "03", "04", "05")))
        } catch (ignore: JSONException) {
            JSONObject()
        }
    }

    internal companion object {
        internal const val FIELD_APP = "app"
        internal const val FIELD_SOURCE = "source"
        internal const val FIELD_FALLBACK_RETURN_URL = "fallback_return_url"

        private const val FIELD_SDK_APP_ID = "sdkAppID"
        private const val FIELD_SDK_TRANS_ID = "sdkTransID"
        private const val FIELD_SDK_ENC_DATA = "sdkEncData"
        private const val FIELD_SDK_EPHEM_PUB_KEY = "sdkEphemPubKey"
        private const val FIELD_SDK_MAX_TIMEOUT = "sdkMaxTimeout"
        private const val FIELD_SDK_REFERENCE_NUMBER = "sdkReferenceNumber"
        private const val FIELD_MESSAGE_VERSION = "messageVersion"
        private const val FIELD_DEVICE_RENDER_OPTIONS = "deviceRenderOptions"

        private const val FIELD_SDK_INTERFACE = "sdkInterface"
        private const val FIELD_SDK_UI_TYPE = "sdkUiType"

        private val MAX_TIMEOUT_FORMATTER = DecimalFormat("00")
    }
}
